manta-pay/src/crypto/hash/poseidon/parameter_generation/matrix.rs

```diff
@@ -0,0 +1,900 @@
+            for j in 0..i {
+                let idx = size - j - 1;
+                shadow_normalized = vec_sub(
+                    &shadow_normalized,
+                    &scalar_vec_mul(&normalized[idx], &shadow_result[j]),
+                );
+                normalized = vec_sub(&normalized, &scalar_vec_mul(&normalized[idx], &result[j]));
+            }
+            result.push(normalized);
+            shadow_result.push(shadow_normalized);
+        }
+        result.reverse();
+        shadow_result.reverse();
+        *shadow = Self(Matrix(shadow_result));
+        Some(Self(Matrix(result)))
+    }
+
+    /// Generates the upper triangular matrix
+    fn upper_triangular(&self, shadow: &mut Self) -> Option<Self>
+    where
+        F: Clone,
+    {
+        let size = self.num_rows();
+        let mut result = Vec::with_capacity(size);
+        let mut shadow_result = Vec::with_capacity(size);
+        let mut current = self.0.clone();
+        let mut shadow_matrix = shadow.0.clone();
+        for column in 0..(size - 1) {
+            current = current.eliminate(column, &mut shadow_matrix)?;
+            result.push(current.0.remove(0));
+            shadow_result.push(shadow_matrix.0.remove(0));
+        }
+        result.push(current.0.take_first());
+        shadow_result.push(shadow_matrix.0.take_first());
+        *shadow = Self(Matrix(shadow_result));
+        Some(Self(Matrix(result)))
+    }
+}
+
+impl<F> AsRef<Matrix<F>> for SquareMatrix<F>
+where
+    F: Field,
+{
+    fn as_ref(&self) -> &Matrix<F> {
+        &self.0
+    }
+}
+
+impl<F> Deref for SquareMatrix<F>
+where
+    F: Field,
+{
+    type Target = Matrix<F>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<F> PartialEq<Matrix<F>> for SquareMatrix<F>
+where
+    F: Field + PartialEq,
+{
+    fn eq(&self, other: &Matrix<F>) -> bool {
+        self.0.eq(other)
+    }
+}
+
+impl<F> MatrixOperations for SquareMatrix<F>
+where
+    F: Field,
+{
+    type Scalar = F;
+
+    fn eliminate(&self, column: usize, shadow: &mut Self) -> Option<Self>
+    where
+        Self::Scalar: Clone,
+    {
+        self.0.eliminate(column, &mut shadow.0).map(Self)
+    }
+
+    fn identity(n: usize) -> Self {
+        Self(Matrix::identity(n))
+    }
+
+    fn matmul(&self, other: &Self) -> Option<Self>
+    where
+        Self::Scalar: Clone,
+    {
+        self.0.matmul(&other.0).map(Self)
+    }
+
+    fn mul_by_scalar(&self, scalar: Self::Scalar) -> Self {
+        Self(self.0.mul_by_scalar(scalar))
+    }
+
+    fn to_row_major(self) -> Vec<F> {
+        self.0.to_row_major()
+    }
+
+    fn transpose(self) -> Self {
+        Self(self.0.transpose())
+    }
+}
+
+/// Inner product of two vectors.
+pub fn inner_product<F>(a: &[F], b: &[F]) -> F
+where
+    F: Field,
+{
+    a.iter()
+        .zip(b)
+        .fold(F::zero(), |acc, (v1, v2)| F::add(&acc, &F::mul(v1, v2)))
+}
+
+/// Elementwise addition (i.e., out_i = a_i + b_i).

```

Comment:
Replace "elementwise" with "pointwise" or just say vector addition.

Use present tense language, try to answer the question "What does this function do?"

Reply:
we should use elementwise instead of pointwise, since `elementwise` is more commonly used: https://www.google.com/search?q=elementwise+addition&oq=elementwise+addition&aqs=chrome..69i57j0i10i512l3j0i512j0i10i512l2j0i22i30l3.2674j0j7&sourceid=chrome&ie=UTF-8